### PR TITLE
Fix progress bar deadlock on DDP metrics computation.

### LINKED ISSUE #21264: TQDMProgressBar bug (?) causes DDP to hang
https://github.com/Lightning-AI/pytorch-lightning/blob/918a1a62635e41b72131fc153920594f0f64cc5b/src/lightning/pytorch/callbacks/progress/tqdm_progress.py#L285-L286

Is this^ a safe implementation for DDP?

Based on my understanding, this will lead to the following stack trace, ultimately leading to a sync operation on `ResultMetrics` if certain conditions are met. However, `self.train_progress_bar.disable` is only `False` (i.e. enabled) on the rank-zero process, so the metrics computations will only succeed if there happens to be another codepath on non-rank-zero devices that's requesting `self._logger_connector.metrics` at the same time.

<details>
  <summary>Stack trace</summary>
  
https://github.com/Lightning-AI/pytorch-lightning/blob/918a1a62635e41b72131fc153920594f0f64cc5b/src/lightning/pytorch/callbacks/progress/progress_bar.py#L180-L201

https://github.com/Lightning-AI/pytorch-lightning/blob/918a1a62635e41b72131fc153920594f0f64cc5b/src/lightning/pytorch/trainer/trainer.py#L1675-L1683

https://github.com/Lightning-AI/pytorch-lightning/blob/918a1a62635e41b72131fc153920594f0f64cc5b/src/lightning/pytorch/trainer/connectors/logger_connector/logger_connector.py#L253-L258

https://github.com/Lightning-AI/pytorch-lightning/blob/918a1a62635e41b72131fc153920594f0f64cc5b/src/lightning/pytorch/trainer/connectors/logger_connector/logger_connector.py#L232-L237

https://github.com/Lightning-AI/pytorch-lightning/blob/918a1a62635e41b72131fc153920594f0f64cc5b/src/lightning/pytorch/trainer/connectors/logger_connector/result.py#L471-L476

https://github.com/Lightning-AI/pytorch-lightning/blob/918a1a62635e41b72131fc153920594f0f64cc5b/src/lightning/pytorch/trainer/connectors/logger_connector/result.py#L425-L440

</details>

I ran into this when debugging an issue in my training script where DDP was hanging at the end of the first epoch. Changing the original code chunk as such resolved the deadlock for me.

```
metrics = self.get_metrics(trainer, pl_module)
if not self.train_progress_bar.disable: 
    self.train_progress_bar.set_postfix(metrics) 
```

### PR DESCRIPTION
**Pull request by tchittesh**
## What does this PR do?

Fixes #21264

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
📚 Documentation preview 📚: https://pytorch-lightning--21322.org.readthedocs.build/en/21322/

### CODE REVIEW

File: tests/tests_pytorch/callbacks/progress/test_rich_progress_bar.py [hunk 1]
```diff
@@ -605,3 +607,55 @@ def val_dataloader(self):
...
     trainer.fit(model)
+
+
+def test_rich_progress_bar_ddp_deadlock(tmp_path):
+    """Tests that RichProgressBar doesn't deadlock when using DDP on train epoch end.
+
+    We used to have a bug where metrics were synced only on the rank 0 process. See
+    https://github.com/Lightning-AI/pytorch-lightning/issues/21264
+    for more details.
+
```
Comment (SkafteNicki) [inline tests/tests_pytorch/callbacks/progress/test_rich_progress_bar.py:618]:
```suggestion
```

File: tests/tests_pytorch/callbacks/progress/test_rich_progress_bar.py [hunk 2]
```diff
@@ -605,3 +607,55 @@ def val_dataloader(self):
...
+    """
+    RichProgressBar()
```
Comment (SkafteNicki) [inline tests/tests_pytorch/callbacks/progress/test_rich_progress_bar.py:620]:
```suggestion
```
seem to not really be used?

File: tests/tests_pytorch/callbacks/progress/test_tqdm_progress_bar.py [hunk 3]
```diff
@@ -859,3 +861,57 @@ def reset(self, total=None):
...
+
+
+@patch("lightning.pytorch.trainer.connectors.callback_connector._RICH_AVAILABLE", False)
+def test_tqdm_progress_bar_ddp_deadlock(tmp_path):
+    """Tests that TQDMProgressBar doesn't deadlock when using DDP on train epoch end.
+
+    We used to have a bug where metrics were synced only on the rank 0 process. See
+    https://github.com/Lightning-AI/pytorch-lightning/issues/21264
+    for more details.
+
```
Comment (SkafteNicki) [inline tests/tests_pytorch/callbacks/progress/test_tqdm_progress_bar.py:873]:
```suggestion
```
